"Spawn Egg item" and "Datagen: Item Models"

Fellteros opened the pull request:
Heya,

I wrote a page on model generation a while ago (for blocks), and I wanted to complete it, so I got together a page on item models. It contains 6 examples - simple, handheld, spawn eggs, conditional, dyeable and custom item models. Some sentences may be a bit wierdly written, since English isn't my native language, so feel free to correct them.

---

integrates and closes #328

Fellteros committed:
init

Fellteros committed:
Fixed conditional item models + minor tweaks

FabricMCBot committed:
New Crowdin updates (#323)

* New translations finding-mods.md (Portuguese, Brazilian)

* New translations linux.md (Russian)

* New translations linux.md (Chinese Simplified)

* New translations linux.md (Chinese Traditional)

* New translations linux.md (Portuguese, Brazilian)

* New translations windows.md (Italian)

* New translations windows.md (Korean)

* New translations windows.md (Polish)

* New translations windows.md (Russian)

* New translations windows.md (Chinese Simplified)

* New translations windows.md (Chinese Traditional)

* New translations installing-mods.md (Chinese Simplified)

* New translations installing-mods.md (Chinese Traditional)

* New translations crash-reports.md (Chinese Simplified)

* New translations uploading-logs.md (Korean)

* New translations uploading-logs.md (Chinese Simplified)

* New translations uploading-logs.md (Chinese Traditional)

* New translations updating-fabric.md (Chinese Simplified)

* New translations custom.md (Chinese Simplified)

* New translations using-sounds.md (Chinese Simplified)

* New translations introduction-to-fabric-and-modding.md (Korean)

* New translations introduction-to-fabric-and-modding.md (Chinese Simplified)

* New translations introduction-to-fabric-and-modding.md (Chinese Traditional)

* New translations launching-the-game.md (Japanese)

* New translations launching-the-game.md (Chinese Simplified)

* New translations launching-the-game.md (Chinese Traditional)

* New translations project-structure.md (Korean)

* New translations project-structure.md (Chinese Simplified)

* New translations custom-armor.md (Italian)

* New translations custom-armor.md (Korean)

* New translations custom-item-groups.md (Chinese Simplified)

* New translations custom-item-interactions.md (Korean)

* New translations custom-tools.md (Chinese Simplified)

* New translations contributing.md (Korean)

* New translations contributing.md (Chinese Simplified)

* New translations blockstates.md (Italian)

* New translations blockstates.md (Chinese Simplified)

* New translations custom-data-components.md (Chinese Simplified)

* New translations custom-enchantment-effects.md (Korean)

* New translations custom-enchantment-effects.md (Chinese Simplified)

* New translations dynamic-sounds.md (Italian)

* New translations block-entity-renderer.md (Chinese Simplified)

* New translations codecs.md (Chinese Traditional)

* New translations sidebar_translations.json (Chinese Traditional)

* New translations website_translations.json (Chinese Traditional)

* New translations website_translations.json (Italian)

* New translations text-and-translations.md (Ukrainian)

* New translations website_translations.json (Ukrainian)

* New translations custom-tools.md (Ukrainian)

* New translations dynamic-sounds.md (Ukrainian)

* New translations index.md (Japanese)

* New translations sidebar_translations.json (Japanese)

* New translations creating-a-project.md (Japanese)

* New translations project-structure.md (Japanese)

* New translations website_translations.json (Japanese)

* New translations setting-up-a-development-environment.md (Japanese)

* New translations setting-up-a-development-environment.md (Russian)

* New translations index.md (Japanese)

* New translations faq.md (Russian)

* New translations networking.md (Ukrainian)

* New translations codecs.md (German)

* New translations codecs.md (Italian)

* New translations codecs.md (Ukrainian)

* New translations codecs.md (Chinese Simplified)

* New translations codecs.md (Chinese Traditional)

* New translations arguments.md (Italian)

* New translations arguments.md (Ukrainian)

* New translations arguments.md (Chinese Simplified)

* New translations basics.md (German)

* New translations basics.md (Italian)

* New translations basics.md (Russian)

* New translations basics.md (Ukrainian)

* New translations basics.md (Chinese Simplified)

* New translations suggestions.md (Italian)

* New translations suggestions.md (Ukrainian)

* New translations suggestions.md (Chinese Simplified)

* New translations effects.md (Italian)

* New translations effects.md (Chinese Simplified)

* New translations events.md (Italian)

* New translations events.md (Ukrainian)

* New translations events.md (Chinese Simplified)

* New translations index.md (German)

* New translations index.md (Italian)

* New translations index.md (Japanese)

* New translations index.md (Ukrainian)

* New translations index.md (Chinese Simplified)

* New translations index.md (Chinese Traditional)

* New translations potions.md (Italian)

* New translations potions.md (Ukrainian)

* New translations potions.md (Chinese Simplified)

* New translations basic-concepts.md (Italian)

* New translations basic-concepts.md (Ukrainian)

* New translations basic-concepts.md (Chinese Simplified)

* New translations draw-context.md (Ukrainian)

* New translations draw-context.md (Chinese Simplified)

* New translations custom-screens.md (Ukrainian)

* New translations custom-widgets.md (German)

* New translations custom-widgets.md (Ukrainian)

* New translations hud.md (German)

* New translations hud.md (Italian)

* New translations hud.md (Ukrainian)

* New translations hud.md (Chinese Simplified)

* New translations creating-particles.md (Italian)

* New translations creating-particles.md (Ukrainian)

* New translations creating-particles.md (Chinese Simplified)

* New translations index.md (Czech)

* New translations faq.md (German)

* New translations faq.md (Ukrainian)

* New translations faq.md (Chinese Simplified)

* New translations finding-mods.md (Ukrainian)

* New translations finding-mods.md (Chinese Simplified)

* New translations index.md (Ukrainian)

* New translations installing-fabric.md (Italian)

* New translations installing-fabric.md (Ukrainian)

* New translations installing-fabric.md (Chinese Simplified)

* New translations linux.md (Italian)

* New translations linux.md (Russian)

* New translations linux.md (Ukrainian)

* New translations linux.md (Chinese Simplified)

* New translations windows.md (German)

* New translations windows.md (Italian)

* New translations windows.md (Russian)

* New translations windows.md (Ukrainian)

* New translations windows.md (Chinese Simplified)

* New translations installing-mods.md (German)

* New translations installing-mods.md (Italian)

* New translations installing-mods.md (Russian)

* New translations installing-mods.md (Ukrainian)

* New translations installing-mods.md (Chinese Simplified)

* New translations crash-reports.md (Ukrainian)

* New translations crash-reports.md (Chinese Simplified)

* New translations crash-reports.md (Chinese Traditional)

* New translations uploading-logs.md (Ukrainian)

* New translations uploading-logs.md (Chinese Simplified)

* New translations updating-fabric.md (German)

* New translations updating-fabric.md (Dutch)

* New translations updating-fabric.md (Ukrainian)

* New translations updating-fabric.md (Chinese Simplified)

* New translations sidebar_translations.json (Czech)

* New translations damage-types.md (Ukrainian)

* New translations damage-types.md (Chinese Simplified)

* New translations custom.md (German)

* New translations custom.md (Italian)

* New translations custom.md (Russian)

* New translations custom.md (Ukrainian)

* New translations custom.md (Chinese Simplified)

* New translations using-sounds.md (Italian)

* New translations using-sounds.md (Russian)

* New translations using-sounds.md (Ukrainian)

* New translations using-sounds.md (Chinese Simplified)

* New translations creating-a-project.md (German)

* New translations creating-a-project.md (Ukrainian)

* New translations creating-a-project.md (Chinese Traditional)

* New translations introduction-to-fabric-and-modding.md (German)

* New translations introduction-to-fabric-and-modding.md (Italian)

* New translations introduction-to-fabric-and-modding.md (Ukrainian)

* New translations introduction-to-fabric-and-modding.md (Chinese Simplified)

* New translations introduction-to-fabric-and-modding.md (Chinese Traditional)

* New translations launching-the-game.md (Ukrainian)

* New translations launching-the-game.md (Chinese Simplified)

* New translations launching-the-game.md (Chinese Traditional)

* New translations project-structure.md (Ukrainian)

* New translations project-structure.md (Chinese Simplified)

* New translations setting-up-a-development-environment.md (German)

* New translations setting-up-a-development-environment.md (Italian)

* New translations setting-up-a-development-environment.md (Russian)

* New translations setting-up-a-development-environment.md (Ukrainian)

* New translations custom-armor.md (Italian)

* New translations custom-armor.md (Russian)

* New translations custom-armor.md (Ukrainian)

* New translations custom-armor.md (Chinese Simplified)

* New translations custom-item-groups.md (Ukrainian)

* New translations custom-item-groups.md (Chinese Simplified)

* New translations custom-item-interactions.md (Ukrainian)

* New translations custom-item-interactions.md (Chinese Simplified)

* New translations custom-tools.md (Ukrainian)

* New translations custom-tools.md (Chinese Simplified)

* New translations first-item.md (German)

* New translations first-item.md (Italian)

* New translations first-item.md (Russian)

* New translations first-item.md (Ukrainian)

* New translations first-item.md (Chinese Simplified)

* New translations first-item.md (Chinese Traditional)

* New translations food.md (Ukrainian)

* New translations text-and-translations.md (German)

* New translations text-and-translations.md (Chinese Simplified)

* New translations contributing.md (German)

* New translations contributing.md (Italian)

* New translations contributing.md (Japanese)

* New translations contributing.md (Ukrainian)

* New translations contributing.md (Chinese Simplified)

* New translations website_translations.json (Czech)

* New translations website_translations.json (German)

* New translations website_translations.json (Italian)

* New translations website_translations.json (Dutch)

* New translations website_translations.json (Ukrainian)

* New translations website_translations.json (Chinese Simplified)

* New translations blockstates.md (Italian)

* New translations blockstates.md (Ukrainian)

* New translations blockstates.md (Chinese Simplified)

* New translations first-block.md (German)

* New translations first-block.md (Italian)

* New translations first-block.md (Ukrainian)

* New translations first-block.md (Chinese Simplified)

* New translations ide-tips-and-tricks.md (German)

* New translations ide-tips-and-tricks.md (Italian)

* New translations ide-tips-and-tricks.md (Ukrainian)

* New translations ide-tips-and-tricks.md (Chinese Simplified)

* New translations custom-data-components.md (Italian)

* New translations custom-data-components.md (Russian)

* New translations custom-data-components.md (Ukrainian)

* New translations custom-data-components.md (Chinese Simplified)

* New translations automatic-testing.md (Chinese Simplified)

* New translations custom-enchantment-effects.md (Chinese Simplified)

* New translations dynamic-sounds.md (German)

* New translations dynamic-sounds.md (Italian)

* New translations dynamic-sounds.md (Ukrainian)

* New translations block-entities.md (German)

* New translations block-entities.md (Italian)

* New translations block-entities.md (Ukrainian)

* New translations advancements.md (German)

* New translations advancements.md (Italian)

* New translations advancements.md (Ukrainian)

* New translations advancements.md (Chinese Simplified)

* New translations loot-tables.md (German)

* New translations loot-tables.md (Chinese Simplified)

* New translations recipes.md (Chinese Simplified)

* New translations setup.md (Italian)

* New translations setup.md (Chinese Simplified)

* New translations tags.md (Chinese Simplified)

* New translations translations.md (Italian)

* New translations translations.md (Chinese Simplified)

* New translations block-entity-renderer.md (Italian)

* New translations block-entity-renderer.md (Ukrainian)

* New translations block-entity-renderer.md (Chinese Simplified)

* New translations block-models.md (German)

* New translations block-models.md (Italian)

* New translations block-models.md (Ukrainian)

* New translations block-models.md (Chinese Simplified)

* New translations fabric-api.md (Chinese Simplified)

* New translations index.md (German)

* New translations index.md (Ukrainian)

* New translations index.md (Chinese Simplified)

* New translations options.md (Chinese Simplified)

* New translations production-run-tasks.md (Ukrainian)

* New translations production-run-tasks.md (Chinese Simplified)

* New translations networking.md (German)

* New translations networking.md (Italian)

* New translations networking.md (Ukrainian)

* New translations networking.md (Chinese Simplified)

Fellteros committed:
Item Model generation page

Fellteros committed:
added textures

Fellteros committed:
lint + checkstyle

VatinMc committed:
Incorporate VatinMc's Spawn Egg article

Co-authored-by: VatinMc <mc@vatin.de>
Signed-off-by: Fellteros <aloisprochazka6@gmail.com>

Fellteros committed:
Added composite, select and range dispatch examples

Signed-off-by: Fellteros <aloisprochazka6@gmail.com>

Fellteros committed:
Added a few screenshots for easier understanding

Signed-off-by: Fellteros <aloisprochazka6@gmail.com>

Fellteros committed:
Markdown lint

Signed-off-by: Fellteros <aloisprochazka6@gmail.com>

Fellteros committed:
Markdown lint again

Signed-off-by: Fellteros <aloisprochazka6@gmail.com>

its-miroma commented:
Hi @Fellteros,

There seems to be a formatting issue in the `Co-authored-by` line of commit `2e53cde`.
Could you please `edit` it?

<details>
<summary>Instructions</summary>

1. Run `git rebase -i main`
2. Change `pick` to `edit` for commit `2e53cde`, and `drop 3a2ba6d` which is empty:
```diff
 pick ab3f59c init
 pick ccd8c7f Fixed conditional item models + minor tweaks
 pick 98dcd2d New Crowdin updates (#323)
 pick c10e658 Item Model generation page
 pick f1d7390 added textures
 pick 8093970 lint + checkstyle
-pick 3a2ba6d lint + checkstyle
+drop 3a2ba6d lint + checkstyle
-pick 2e53cde Incorporated VatinMc's Spawn Egg article. [...]
+edit 2e53cde Incorporated VatinMc's Spawn Egg article. [...]
 pick 2b43088 Added composite, select and range dispatch examples
 pick 4fb19e5 Markdown lint
 pick 4fe469d Added a few screenshots for easier understanding
 pick 167167a Markdown lint again
```
3. Save and close the file
4. The `rebase` should stop at `2e53cde`. Run `git commit --amend --author="VatinMc <mc@vatin.de>"`
5. A `COMMIT_EDITMSG` file should open, allowing you to fix the message. Please put an empty line between the title and the `Co-authored-by` line:
```md
Incorporate VatinMc's Spawn Egg article

Co-authored-by: VatinMc <mc@vatin.de>
Signed-off-by: Fellteros <aloisprochazka6@gmail.com>
```
6. Save and close the file
7. Run `git rebase --continue`
8. Run `git push --force`

</details>

Fellteros commented:
@its-miroma Done

Fellteros committed:
Merge branch 'main' into item-model-generation

Fellteros committed:
Added info about the inner static class CustomItemModelGenerator

Signed-off-by: Fellteros <aloisprochazka6@gmail.com>

Fellteros committed:
Merge remote-tracking branch 'origin/item-model-generation' into item-model-generation

Fellteros committed:
Merge branch 'main' into item-model-generation

skycatminepokie commented on .vitepress/sidebars/develop.ts:
```diff
@@ -92,6 +92,10 @@ export default [
       },
+      {
+        text: "develop.items.spawn-egg",
+        link: "/develop/items/spawn-egg",
```
Perhaps the page should go under "entities?" That's where I'd look for it first.

skycatminepokie commented on develop/items/spawn-egg.md:
```diff
@@ -0,0 +1,103 @@
+Spawn eggs are items that are used to spawn the corresponding mob. When creating spawn eggs, the following questions need to be answered:
+
+- Which `MobEntity` should be spawned?
+- What colors should shell and dots be?
```
```suggestion
- What colors should the shell and dots be?
```

skycatminepokie commented on develop/data-generation/item-models.md:
```diff
@@ -0,0 +1,247 @@
+<DownloadEntry visualURL="/assets/develop/data-generation/item-model/flashlight_textures_big.png" downloadURL="/assets/develop/data-generation/item-model/flashlight_textures.zip">Flashlight</DownloadEntry>
+
+:::warning IMPORTANT
+When obtaining `Identifier`s for the `ItemModels.basic()`, always use `itemModelGenerator.upload()` or any other method using it, otherwise your item model files won't generate, only item model definitions!
```
I don't understand this line.

Fellteros replied:
If you'd pass in only an `Identifier`, the model wouldn't be added to the `modelCollector`, which is essentially a supplier of all models for which JSON files should be generated.

skycatminepokie commented on develop/data-generation/item-models.md:
```diff
@@ -0,0 +1,247 @@
+
+@[code lang=java transcludeWith=:::datagen-model:composite](@/reference/latest/src/client/java/com/example/docs/datagen/FabricDocsReferenceModelProvider.java)
+
+These item models are composed of one or more textures layered on top of each other. There isn't any vanilla method for this, you have to use `ItemModelGenerator`'s `output` field and then `accept()` for it to be working.
```
```suggestion
These item models are composed of one or more textures layered on top of each other. There aren't any vanilla methods for this, you have to use `ItemModelGenerator`'s `output` field and then `accept()` for it to work.
```

skycatminepokie commented on develop/data-generation/item-models.md:
```diff
@@ -0,0 +1,247 @@
+| `DisplayContextProperty`    | Renders an item model based on the position in which the item is (`ground`, `fixed`, `head`, ...). |
+| `ContextEntityTypeProperty` | Renders an item model based on the entity holding the item.                                        |
+
+In this example, the item changes texture when traveling between dimensions; green in Overworld, red in the Nether and black in the End.
```
TIL you can do that

skycatminepokie commented on develop/data-generation/item-models.md:
```diff
@@ -0,0 +1,247 @@
+
+@[code lang=java transcludeWith=:::datagen-model:range-dispatch](@/reference/latest/src/client/java/com/example/docs/datagen/FabricDocsReferenceModelProvider.java)
+
+Renders an item model based on the value of a numeric property. Take in an item and list of variants paired with a value. There are quite a few, here are some examples:
```
Perhaps info on where to find more? Is there a common interface? Not necessary, just a minor touch-up.

skycatminepokie commented on develop/data-generation/item-models.md:
```diff
@@ -0,0 +1,247 @@
+1. ``Item item``: The item, for which we are generating the models (in this case `ModItems.BALLOON`).
+2. ``ItemModelGenerator generator``: the same that get passed into the `generateItemModels()` method. Used for its fields.
+
+First, we get the ``Identifier`` of the item with `BALLOON.upload()`, passing in a ``TextureMap`` and the `modelCollector` from our `generator` parameter.
```
I'm curious, are the double graves (\`\`) for yarn-mapped things?

skycatminepokie replied:
Reviewed as a dev knowing nothing about item model definitions.

its-miroma replied:
> I'm curious, are the double graves (\`\`) for yarn-mapped things?

no, they are unnecessary.

Fellteros committed:
Refactor according to skycatminepokie's feedback

Co-authored-by: skycatminepokie <52087591+skycatminepokie@users.noreply.github.com>

cassiancc committed:
Merge branch 'main' into pr/329

cassiancc committed:
Port to 1.21.10 and Mojang Mappings

Spawn egg article will need heavy reworks - they're pretty much just normal items now.

cassiancc commented:
Dealt with the merge conflicts here. ~~Now that we're working with 1.21.10, it's worth considering whether the spawn egg article is needed since they're effectively just normal items with an extra component. Not opposed to it, will just need rewording + a new texture.~~
Gone ahead and updated this page. If food can have a dedicated page, so can Spawn Eggs.

cassiancc committed:
Rework spawn egg info for 1.21.10

cassiancc reviewed (dismissed):
This is good to go IMO. Once it's merged, we can start integrating it with the custom entity tutorial.

cassiancc committed:
Merge branch 'main' into item-models

cassiancc committed:
move spawn egg info into entities

cassiancc committed:
Revert "move spawn egg info into entities"

This reverts commit 0bc6cf0f551c97b121e9306d087ea1d602efb15d.

cassiancc committed:
spawn eggs are items i suppose

its-miroma committed:
drop changes to 1.21.4

cassiancc committed:
Fix fabric docs reference

cassiancc committed:
simplify spawn egg registration

cassiancc committed:
checkstyle

cassiancc committed:
fix mappings on block model page

cassiancc committed:
Update ModItems.java

cassiancc committed:
Link directly to `generated.json`

cassiancc committed:
Fix sidebar

its-miroma committed:
my review

- drop Troubleshooting section of Spawn Eggs
- shorter headings
- use tabs for: source code, client item, item model, texture, preview
- show generated item models for all built-in methods
- use hex format for color
- have all `DownloadEntry`s say "Download ... Texture(s)`
- avoid `datagen-model:` and `datagen-model-custom:` prefixes
- don't show `Custom*ModelGenerator` in `:::provider`, show it alone
- use properties when referring to `Item.Properties`
- Uniform static references in `ModBlocks`
- Avoid redundant `lang=java` in transclusion

cassiancc committed:
fix mapping issues

its-miroma committed:
format sidebar

its-miroma committed:
more of mojmap migration

skycatminepokie commented on develop/data-generation/item-models.md:
```diff
@@ -0,0 +1,373 @@
+---
+title: Item Model Generation
+description: A guide to generating item models via datagen.
+authors:
```
Does this need to be alphabetized?

skycatminepokie commented on develop/data-generation/item-models.md:
```diff
@@ -0,0 +1,373 @@
+
+### Handheld {#handheld}
+
+Handheld item models are generally used by tools and weapons (axes, swords, trident). They are rotated and positioned a little differently from the simple models to look more natural in hand. Other than that, they look exactly the same as the simple ones.
```
```suggestion
Handheld item models are generally used by tools and weapons (axes, swords, trident). They are rotated and positioned a little differently from the simple models to look more natural in hand.
```
Could shorten if desired.

skycatminepokie commented on develop/data-generation/item-models.md:
```diff
@@ -0,0 +1,373 @@
+
+### Dyeable {#dyeable}
+
+Dyeable items require a default decimal color value, which is used when the item is not dyed. The default value for leather is `0xFFA06540`. The method for dyeable items generates a simple item model and a client item which specifies the tint color.
```
Tint color - is that default color?

skycatminepokie replied:
Or a tint after adding dye?

cassiancc replied:
I reorganized this paragraph to hopefully make it more clear.

skycatminepokie commented on develop/items/spawn-egg.md:
```diff
@@ -0,0 +1,63 @@
+---
+title: Spawn Eggs
+description: Learn how to register a spawn egg item.
+authors:
```
Does this need to be alphabetized?

skycatminepokie commented on develop/items/spawn-egg.md:
```diff
@@ -0,0 +1,63 @@
+You must first understand [how to create an item](./first-item), which you can then turn into a spawn egg.
+:::
+
+Spawn eggs are special items that, when used, spawn their corresponding mob. You can register one with the `register` method from your [item class](./first-item#preparing-your-items-class), by passing `SpawnEggItem::new` to it.
```
```suggestion
Spawn eggs are special items that, when used, spawn their corresponding mob. You can register one with the `register` method from your [items class](./first-item#preparing-your-items-class), by passing `SpawnEggItem::new` to it.
```

skycatminepokie commented on develop/items/spawn-egg.md:
```diff
@@ -0,0 +1,63 @@
+
+## Adding To A Creative Mode Tab {#adding-to-a-creative-mode-tab}
+
+The spawn egg is added to the spawn egg `CreativeModeTab` in the `initialize()` method of the [item class](./first-item#preparing-your-items-class).
```
```suggestion
The spawn egg is added to the spawn egg `CreativeModeTab` in the `initialize()` method of the [items class](./first-item#preparing-your-items-class).
```

skycatminepokie reviewed (approved):
Only nitpicks, so approved.

cassiancc committed:
Apply suggestions from code review

Co-authored-by: skycatminepokie <52087591+skycatminepokie@users.noreply.github.com>

cassiancc committed:
Alphabetize authors

cassiancc committed:
Review comments

its-miroma commented:
@cassiancc dicedpixels on Discord reminded us that the spawn egg screenshots were not up to date with the example-mod migration.

cassiancc commented:
Ah, that's my bad, didn't see anything on the PR itself. Happy to approve a PR to fix that.
